email/fetcher.go

```diff
@@ -0,0 +1,393 @@
+package email
+
+import (
+	"abuse-scanner/database"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+
+	"github.com/emersion/go-imap"
+	"github.com/emersion/go-imap/client"
+)
+
+const (
+	// fetchFrequency defines the frequency with which we fetch new emails
+	fetchFrequency = 30 * time.Second
+)
+
+type (
+	// Fetcher is an object that will periodically scan an inbox and persist the
+	// missing messages in the database.
+	Fetcher struct {
+		staticContext     context.Context
+		staticDatabase    *database.AbuseScannerDB
+		staticEmailClient *client.Client
+		staticLogger      *logrus.Entry
+		staticMailbox     string
+	}
+
+	// messageIDMap is a helper struct that maps the message uid to its seq num
+	messageIDMap map[uint32]uint32
+)
+
+// NewFetcher creates a new fetcher.
+func NewFetcher(ctx context.Context, database *database.AbuseScannerDB, emailClient *client.Client, mailbox string, logger *logrus.Logger) *Fetcher {
+	return &Fetcher{
+		staticContext:     ctx,
+		staticDatabase:    database,
+		staticEmailClient: emailClient,
+		staticLogger:      logger.WithField("module", "Fetcher"),
+		staticMailbox:     mailbox,
+	}
+}
+
+// Start initializes the fetch process.
+func (f *Fetcher) Start() error {
+	// list mailboxes
+	mailboxes, err := f.listMailboxes()
+	if err != nil {
+		return err
+	}
+
+	// check whether the mailbox exists
+	found := false
+	for _, m := range mailboxes {
+		if m == f.staticMailbox {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return errors.New("mailbox not found")
+	}
+
+	go f.threadedFetchMessages()
+	return nil
+}
+
+// listMailboxes returns all mailboxes
```

Comment:
It's not entirely clear to me what a mailbox is in this context. Is that a specific email address (e.g. `noreply@siasky.net`) that belongs to an email account that holds several of those? It would be helpful to expand the docstring for people like me who don't have a lot of the required context.